keps/sig-node/20191211-node-maintenance-lease.md

```diff
@@ -0,0 +1,175 @@
+
+Utilize the existing `Lease` built-in API in the API group `coordination.k8s.io`.
+Create a new Lease object per-node  with Name equal to Node name in a newly
+created dedicated namespace
+That namespace should be created automatically (similarly to "default" and
+"kube-system", probably by NodeController) and never be deleted (so that nodes
+don't require permission for it).
+
+The Lease object should be created automatically, and the ownerRef should
+be the corresponding node so it is removed when the node is deleted.
+
+### User Stories [optional]
+
+#### Story 1
+
+I have a configuration management system that requires the node to be online to perform configuration updates.
+
+I have an automated-snapshot system that stops an instance and takes a snapshot for backup / data retention purposes.
+
+I do not want snapshots to happen during an update.
+
+#### Story 2
+
+I am investigating a problem with the node. I don't want the node to be disrupted
+by automation while I'm performing my investigation or fixes, and I don't want
+to stop the automation components from acting on other nodes.
+
+### Implementation Details/Notes/Constraints
+
+#### Impact to the Kubelet
+
+None. The kubelet does not need to respond to this lease or interact with it in
+any way.  This is for other components to coordination actions they intend to
+take (or prevent) on the node that might be disruptive.
+
+#### Impact to Scheduleability
+
+None. The existence of acquiring of this lease does not imply any taints,
+annotations, labels, cordoning or draining will take place. The Lease object
+itself is not intended as a mechanism to start any automated process.
+
+A user/controller acquires the Lease object, and then that user/controller
+may add/remove taints, cordon or drain, if they choose to.  Other components
+may wish to integrate with the new Lease object to avoid conflict. EG:
+a component that ensures a taint is always present on a node might stop ensuring
+that taint if there is a current valid lease holder, but that implementation
+is out of scope here.
+
+#### Client Behavior
+
+1. Client has required RBAC permissions for new namespace and Lease object
+1. Client checks for the existence of a lease (to be backwards compatible)
+1. If the lease exists, confirm the lease is not currently held by another user
+1. Attempt to acquire the lease and set an appropriate LeaseDurationSeconds
+1. If acquired, perform the necessary operations
+1. Release the lease
+
+### Risks and Mitigations
+
+Clients might not behave appropriately.  We can mitigate this by providing
+a library with an easy to use interface.
+
+## Alternatives
+
+### Annotations on Node or elsewhere
+
+Use a designated annotation on the node object, or possibly
+elsewhere.  Drawbacks to this include excess updates being applied to the node
+object directly, which might have performance implications. Having such an
+annotation natively seems like an anti-pattern and could easily be disrupted
```

Comment:
I'm reading this statement as saying that we can't use annotations as a locking mechanism, as that is not what they were designed for - I'm I understanding correctly?

Reply:
We could use annotations for this.  I'm not sure this type of operation fits the mold of annotations, and annotations seems to be not meant for this type of coordination; it doesn't seem like there should be a system-wide annotation, I don't think I've seen that.

In any case, if other people decide annotation is the best way/place to do this, it's perfectly fine for me, but the drawback of many things reading/writing to the node makes it less than ideal IMO.

Reply:
If you used annotations for this, what would that look like, a key with a value which is the lease end time? What prevents someone from overwriting it? (Is that misbehaving?)

Reply:
Annotations would look pretty ugly, probably.  I haven't sat down and actually drawn out how it would work, but probably it takes more than 1 annotation, possibly 3 or more.  1 for acquire time.  1 for duration/release-time, 1 for ownership.  These values could be coerced into a CSV string, but that's probably even worse.

I think if anyone has a solid implementation alternative they want specified here, I can include it, but I don't feel inclined to specify in such detail something that's not really being considered.